datafusion/expr/src/signature.rs

```diff
@@ -92,14 +92,22 @@ pub enum TypeSignature {
     /// A function such as `concat` is `Variadic(vec![DataType::Utf8, DataType::LargeUtf8])`
     Variadic(Vec<DataType>),
     /// One or more arguments of an arbitrary but equal type.
-    /// DataFusion attempts to coerce all argument types to match the first argument's type
+    /// DataFusion attempts to coerce all argument types to match to the common type with comparision coercion.
     ///
     /// # Examples
     /// Given types in signature should be coercible to the same final type.
     /// A function such as `make_array` is `VariadicEqual`.
     ///
     /// `make_array(i32, i64) -> make_array(i64, i64)`
     VariadicEqual,
+    /// One or more arguments of an arbitrary but equal type or Null.
+    /// Non-comparison coercion is attempted to match the signatures.
+    ///
+    /// Functions like `coalesce` is `VariadicEqual`.
+    // TODO: Temporary Signature, to differentiate existing VariadicEqual.
+    // After we swtich `make_array` to VariadicEqualOrNull,
+    // we can reuse VariadicEqual.
+    VariadicEqualOrNull,
```

Comment:
I think we need a similar signature but an exact args number for `nullif` and `nvl`. 
Can name it as `UniformEqual` compare to `VariadicEqual` 🤔

Reply:
Maybe we could do something like this (basically to flavor the type signature) 🤔 

```rust
pub enum TypeSignature {
...
  /// Rather than the usual coercion rules, special type union rules are applied
  Union(Box<TypeSignature>)
}
```

Reply:
Nice idea!